### PR TITLE
Specify System::PreUpdate, Update execution order

### PR DESCRIPTION
**Pull request by scpeters**
# 🎉 New feature

Part of https://github.com/gazebosim/gz-sim/issues/2268 and https://github.com/gazebosim/gz-sim/issues/2391, second attempt at https://github.com/gazebosim/gz-sim/pull/2394

## Summary

While the `PreUpdate`, `Update`, and `PostUpdate` phases of gz-sim systems allow some control over the order in which code is executed, there are cases in which more control is desired. For example, as noted in #2268 and https://github.com/gazebosim/gz-sim/issues/2391, sensor systems currently update their data during `PostUpdate` (after the `Update` step of the physics system) but can't write their data to the ECM since `PostUpdate` has read-only access to the ECM. Moving sensor updates to `Update` would allow writing to the ECM, but would require ensuring that the sensor systems execute their `Update` step after the physics `Update`. Alternatively, if one wanted to implement a hierarchical cascade of systems acting on data sequentially, rather than adding `PrePreUpdate` and `PrePrePreUpdate` phases, one could specify the system priority during `PreUpdate` for the relevant systems instead.

This feature is used by specifying an integer value in a `<gz:system_priority>` tag for a system that will control the order in which System `PreUpdate` and `Update` callbacks are executed, with smaller priority values executing first.

## Test it

Follow the README instructions to build and run the `PriorityPrinter` example plugin to illustrate the feature.

## Checklist
- [X] Signed all commits for DCO
- [ ] Added tests
- [X] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [X] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [X] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: examples/plugin/priority_printer_plugin/PriorityPrinter.cc [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    gz::sim::System,
+    priority_printer::PriorityPrinter::ISystemConfigure,
+    priority_printer::PriorityPrinter::ISystemPreUpdate,
+    priority_printer::PriorityPrinter::ISystemUpdate)
+
+using namespace priority_printer;
+
+void PriorityPrinter::Configure(
+        const gz::sim::Entity &_entity,
+        const std::shared_ptr<const sdf::Element> &_sdf,
```
Comment (ahcorde) [inline examples/plugin/priority_printer_plugin/PriorityPrinter.cc:43]: include `<memory>`

File: examples/plugin/priority_printer_plugin/PriorityPrinter.hh [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  // this system's priority value from the XML configuration,
+  // and a custom label from the XML configuration during the Update callback.
+  class PriorityPrinter:
+    public gz::sim::System,
+    public gz::sim::ISystemConfigure,
+    public gz::sim::ISystemPreUpdate,
+    public gz::sim::ISystemUpdate
+  {
+    public: void Configure(const gz::sim::Entity &_entity,
+                 const std::shared_ptr<const sdf::Element> &_sdf,
```
Comment (ahcorde) [inline examples/plugin/priority_printer_plugin/PriorityPrinter.hh:36]: include `<memory>`

File: include/gz/sim/System.hh [hunk 3]
```diff
@@ -74,6 +82,19 @@ namespace gz
     /// simulation is started un-paused.
     class System
     {
+      /// \brief Signed integer type used for specifying priority of the
+      /// execution order of PreUpdate and Update phases.
+      public: using PriorityType = int32_t;
```
Comment (ahcorde) [inline include/gz/sim/System.hh:87]: include `<cstdint>`

File: src/SystemManager.cc [hunk 4]
```diff
@@ -109,6 +110,15 @@ size_t SystemManager::ActivatePendingSystems()
   {
     this->systems.push_back(system);
 
+    PriorityType p {System::kDefaultPriority};
+    const std::string kPriorityElementName
```
Comment (ahcorde) [inline src/SystemManager.cc:114]: include `<string>`

Comment (scpeters) [reply src/SystemManager.cc:115]: https://github.com/gazebosim/gz-sim/pull/2487/commits/21f80aed20531852c90d79696e6e7ca7ad16cbad

File: include/gz/sim/System.hh (hunk 3 above)
Comment (scpeters) [reply include/gz/sim/System.hh:88]: https://github.com/gazebosim/gz-sim/pull/2487/commits/21f80aed20531852c90d79696e6e7ca7ad16cbad

File: examples/plugin/priority_printer_plugin/PriorityPrinter.hh (hunk 2 above)
Comment (scpeters) [reply examples/plugin/priority_printer_plugin/PriorityPrinter.hh:37]: https://github.com/gazebosim/gz-sim/pull/2487/commits/21f80aed20531852c90d79696e6e7ca7ad16cbad

File: examples/plugin/priority_printer_plugin/PriorityPrinter.cc (hunk 1 above)
Comment (scpeters) [reply examples/plugin/priority_printer_plugin/PriorityPrinter.cc:44]: https://github.com/gazebosim/gz-sim/pull/2487/commits/21f80aed20531852c90d79696e6e7ca7ad16cbad

**Review by iche033 — COMMENTED** [review]
tested running example world and works as intended. Just a few minor comments.

File: examples/plugin/priority_printer_plugin/README.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
+# Priority Printer
+
+This example illustrates how to control the order of execution of System
+Update callbacks.
```
[the remark this answers is not in the document]
Comment (scpeters) [reply examples/plugin/priority_printer_plugin/README.md]: https://github.com/gazebosim/gz-sim/pull/2487/commits/98cd1be5267ff87c32560c55d744f8471e45a4a0

File: examples/plugin/priority_printer_plugin/README.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+`priority_printer_plugin.sdf` file that's going to be loaded.
+
+Before starting Gazebo, we must make sure it can find the plugin by doing:
+
+~~~
+cd gz-sim/examples/plugins/priority_printer
+export GZ_SIM_SYSTEM_PLUGIN_PATH=`pwd`/build
+~~~
+
+Then load the example world:
```
[the remark this answers is not in the document]
Comment (scpeters) [reply examples/plugin/priority_printer_plugin/README.md]: https://github.com/gazebosim/gz-sim/pull/2487/commits/98cd1be5267ff87c32560c55d744f8471e45a4a0

File: examples/plugin/priority_printer_plugin/README.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+[Msg] PreUpdate: Iteration 5, system priority 100, system label first
+[Msg] PreUpdate: Iteration 5, system priority 100, system label seventh
+[Msg] Update: Iteration 5, system priority -100, system label fifth
+[Msg] Update: Iteration 5, system priority -10, system label third
+[Msg] Update: Iteration 5, system priority unset, system label fourth
+[Msg] Update: Iteration 5, system priority 0, system label sixth
+[Msg] Update: Iteration 5, system priority 10, system label second
+[Msg] Update: Iteration 5, system priority 100, system label first
+[Msg] Update: Iteration 5, system priority 100, system label seventh
+```
```
[the remark this answers is not in the document]
Comment (scpeters) [reply examples/plugin/priority_printer_plugin/README.md:110]: ok, I'll add this

[the remark this answers is not in the document]
Comment (scpeters) [reply examples/plugin/priority_printer_plugin/README.md:128]: I improved the description in https://github.com/gazebosim/gz-sim/pull/2487/commits/d4b161b56444c059ebe497b4e44271f1952474a6

**Review by iche033 — APPROVED** [review]

**Pull request merged**